source/extensions/filters/http/dynamic_modules/filter.cc

```diff
@@ -114,6 +124,100 @@ void DynamicModuleHttpFilter::sendLocalReply(
 
 void DynamicModuleHttpFilter::encodeComplete() {};
 
+bool DynamicModuleHttpFilter::sendHttpCallout(uint32_t callout_id, absl::string_view cluster_name,
+                                              Http::RequestMessagePtr&& message,
+                                              uint64_t timeout_milliseconds) {
+  Upstream::ThreadLocalCluster* cluster =
+      config_->cluster_manager_.getThreadLocalCluster(cluster_name);
+  if (!cluster) {
+    ENVOY_LOG(error, "Cluster {} not found", cluster_name);
+    return false;
+  }
+  Http::AsyncClient::RequestOptions options;
+  options.setTimeout(std::chrono::milliseconds(timeout_milliseconds));
+  auto [iterator, inserted] = http_callouts_.try_emplace(
+      callout_id, std::make_unique<DynamicModuleHttpFilter::HttpCalloutCallback>(shared_from_this(),
+                                                                                 callout_id));
+  if (!inserted) {
+    ENVOY_LOG(error, "Duplicate callout id {}", callout_id);
+    return false;
+  }
+  DynamicModuleHttpFilter::HttpCalloutCallback& callback = *iterator->second;
+  auto request = cluster->httpAsyncClient().send(std::move(message), callback, options);
+  if (!request) {
+    ENVOY_LOG(error, "Failed to send HTTP callout");
```

Comment:
I can't remember but I think this one can fail due to legitimate runtime issues like no healthy upstream? I don't think you should log spam on this one. You need to either use a backoff log, metric, etc. The ones above seem like general config/programming errors to me so I guess OK.

Reply:
Thanks, that's a valid concern. I changed the ABI to return the status instead of the status so that it doesn't need to log at all in the first place.  https://github.com/envoyproxy/envoy/pull/39151/commits/a739f728e874f0ebc718c6660ec3bc28d12dac5e

(i think we should switch to this status style elsewhere as well but i can take a shot at it in follow ups)